File: 백준/Gold/17135. 캐슬 디펜스/캐슬 디펜스.java


import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		n = sc.nextInt();
		m = sc.nextInt();

		d = sc.nextInt();

		arr = new int[m];
		sel = new int[3];
		for (int i = 0; i < m; i++) {
			arr[i] = i;
		}

		xlist_origin = new LinkedList<>();
		ylist_origin = new LinkedList<>();

		int[][] grid = new int[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				grid[i][j] = sc.nextInt();
			}
		}

		for (int j = 0; j < m; j++) {
			for (int i = n - 1; i >= 0; i--) {
				if (grid[i][j] == 1) {
					xlist_origin.add(i);
					ylist_origin.add(j);
				} else {
					continue;
				}
			}
		}
		enumynum = xlist_origin.size();
//		System.out.println(xlist_origin);
//		System.out.println(ylist_origin);

		ans = 0;
		hanzo(0, 0); // 궁수 구하는 조합
		System.out.println(ans);

	}

	static int ans;
	static int[] arr;
	static int[] sel;
	static int n;
	static int m;
	static int d;
	static int enumynum;
	static List<Integer> xlist_origin;
	static List<Integer> ylist_origin;

	private static void hanzo(int idx, int sidx) {
		if (sidx == 3) {
//			System.out.println(Arrays.toString(sel));
			List<Integer> xlist = new LinkedList<>(xlist_origin);
			List<Integer> ylist = new LinkedList<>(ylist_origin);

			int kill = 0;

			gameout: while (true) {
				// 가장 거리가 가까운 거리 먼저 담고 // 그 다음에 좌표 담기.

				if (xlist.isEmpty()) {
//					System.out.println(kill);
					break gameout;
				}

				if (kill == enumynum) {
					break gameout;
				}

				int[] dists = new int[3];
				for (int i = 0; i < 3; i++) {
					int dist = Integer.MAX_VALUE;
					for (int j = 0; j < xlist.size(); j++) { // 적의 수
						int tmp = Math.abs(n - xlist.get(j)) + Math.abs(sel[i] - ylist.get(j));
						if (dist > tmp) {
							dist = tmp;
							if (dist <= d) { // 최대 사격 거리보다 작거나 같으면 죽일 예정
								dists[i] = tmp;
							} else {
								dists[i] = 100; // 아님 못죽임.
							}
						}
					}
				}
				// 죽일 후보 위치 담는 set 생성.
				Set<Integer> killx = new LinkedHashSet<>();

				con: for (int i = 0; i < 3; i++) {
					for (int j = 0; j < xlist.size(); j++) {
						if (dists[i] == Math.abs(n - xlist.get(j)) + Math.abs(sel[i] - ylist.get(j))) {
							killx.add(j);
							continue con;
						}
					}
				}

				kill += killx.size();

				List<Integer> tmpxlist = new LinkedList<>();
				List<Integer> tmpylist = new LinkedList<>();

				for (int j = 0; j < xlist.size(); j++) { // 여기가 꼬일 것 같은데 확인.
					if (!killx.contains(j)) {
						tmpxlist.add(xlist.get(j));
						tmpylist.add(ylist.get(j));
					}
				}

				xlist = tmpxlist;
				ylist = tmpylist;

				// 공격 끝. 내려오기
				for (int i = 0; i < xlist.size(); i++) {
					xlist.set(i, xlist.get(i) + 1);

				}
				// 침략하기

				for (int i = 0; i < xlist.size(); i++) {
					if (xlist.get(i) == n) {
						xlist.remove(i);
						ylist.remove(i);
						i--;

					}
				}

			}
			ans = Math.max(ans, kill);
			return;
		}
		if (idx == m) {
			return;
		}

		sel[sidx] = arr[idx];
		hanzo(idx + 1, sidx + 1);
		hanzo(idx + 1, sidx);
	}
}

File: 백준/Gold/17135. 캐슬 디펜스/캐슬 디펜스.py
'''
궁수 3명
공격은 D이하에서 가장 가까운, 왼쪽
적은 아래로 한 칸이동
모든 적이 격자판에서 제외되면 게잉ㅁ끝
입력
적 1

구상 n 중에 3개를 골름(조합)
궁수를 위치시키고 시뮬레이션 진행
적을 많이 제거해야함
# 패딩 굳이 안주겠다
'''
import heapq

n,m,limit = map(int, input().split())
grid = [list(map(int, input().split())) for i in range(n)]
sel = [0]*3
total_enemy = 0
for _ in grid:
    total_enemy += _.count(1)
ans = 0
def combi(sidx,idx):
    global ans, grid
    if sidx == 3:
        origin_grid = [_[:] for _ in grid]
        cur_enemy = total_enemy
        kill = 0
        while True:
            # 궁수들이 각 후보를 고름
            enemy_list = []
            for lo in sel:
                ele_enemy = []
                for i in range(n):
                    for j in range(m):
                        if grid[i][j] == 1:
                            if (n-i)+abs(lo-j) <= limit:
                                heapq.heappush(ele_enemy,((n-i)+abs(lo-j),j,i))
                if ele_enemy:
                    dist,c,r = heapq.heappop(ele_enemy)
                    enemy_list.append((r,c))

            # 후보 킬 -> 1에서 0 이됐으면 score+1
            for r, c in enemy_list:
                if grid[r][c] == 1:
                    grid[r][c] = 0
                    kill +=1
                    cur_enemy -=1
            # 궁수들 당기는데
            # 맨 밑에애들은 die
            for j in range(m):
                if grid[n-1][j] == 1:
                    grid[n-1][j] = 0
                    cur_enemy -=1

            for i in range(n-1,0,-1):
                for j in range(m):
                    if grid[i-1][j] == 1:
                        grid[i][j],grid[i-1][j] = grid[i-1][j] ,grid[i][j]
            # 적이 0명이면 끝
            if cur_enemy == 0:
                ans = max(ans, kill)
                grid = [_[:] for _ in origin_grid]
                return

    if idx == m:
        return

    sel[sidx] = idx
    combi(sidx+1, idx+1)
    combi(sidx, idx+1)


combi(0,0)
print(ans)